File: src/openrgb_flowers/effects/blooming_engine.py
"""High-performance floral blooming simulation engine."""
from __future__ import annotations
import math
import random
from typing import List, Optional
import numpy as np

from openrgb_flowers.core.interfaces.i_effect_engine import IEffectEngine
from openrgb_flowers.core.interfaces.i_layout_provider import ILayoutProvider
from openrgb_flowers.core.interfaces.i_blend_strategy import IBlendStrategy
from openrgb_flowers.core.interfaces.i_color_palette import IColorPalette
from openrgb_flowers.core.models.effect_config import EffectConfig
from openrgb_flowers.core.models.render_frame import RenderFrame
from openrgb_flowers.effects.flower_instance import FlowerInstance
from openrgb_flowers.effects.palettes.palette_registry import PaletteRegistry
from openrgb_flowers.effects.blending.organic_weighted_blend import OrganicWeightedBlend
from openrgb_flowers.effects.blending.screen_blend import ScreenBlend
from openrgb_flowers.effects.blending.additive_blend import AdditiveBlend
from openrgb_flowers.math.fast_color import FastColorMath
from openrgb_flowers.math.spatial_grid import SpatialGrid


class BloomingEngine(IEffectEngine):
    """Vectorized floral engine orchestrating concurrent blooming flowers."""

    def __init__(
        self,
        config: EffectConfig,
        layout_provider: ILayoutProvider,
        blend_strategy: Optional[IBlendStrategy] = None,
        palette: Optional[IColorPalette] = None,
        spatial_grid: Optional[SpatialGrid] = None,
    ) -> None:
        self._config = config
        self._layout = layout_provider
        self._palette = palette or PaletteRegistry.get(config.palette_name)
        self._blend_strategy = blend_strategy or self._resolve_blend_strategy(config.blend_mode)
        self._spatial_grid = spatial_grid or SpatialGrid()

        # Cache layout coordinates as float arrays
        self._x_coords, self._y_coords = self._layout.get_coordinate_arrays()
        self._led_count = self._layout.get_key_count()

        # Simulation state
        self._flowers: List[FlowerInstance] = []
        self._frame_count = 0
        self._sim_time = 0.0
        self._spawn_accumulator = 0.0

        # Precompute background base array
        self._bg_base = np.tile(
            np.array(self._config.background_color.to_tuple(), dtype=np.float32),
            (self._led_count, 1),
        )

    @staticmethod
    def _resolve_blend_strategy(mode_name: str) -> IBlendStrategy:
        name = mode_name.lower().strip()
        if name == "screen":
            return ScreenBlend()
        elif name == "additive":
            return AdditiveBlend()
        return OrganicWeightedBlend()

    def get_config(self) -> EffectConfig:
        return self._config

    def get_active_flower_count(self) -> int:
        return len(self._flowers)

    def update_layout(self, layout_provider: ILayoutProvider) -> None:
        """Updates the physical layout provider and recomputes coordinate buffers."""
        self._layout = layout_provider
        self._x_coords, self._y_coords = self._layout.get_coordinate_arrays()
        self._led_count = self._layout.get_key_count()
        self._bg_base = np.tile(
            np.array(self._config.background_color.to_tuple(), dtype=np.float32),
            (self._led_count, 1),
        )

    def reset(self) -> None:
        self._flowers.clear()
        self._frame_count = 0
        self._sim_time = 0.0
        self._spawn_accumulator = 0.0

    def _should_spawn_flower(self, dt: float) -> bool:
        """Determines if a new flower blossom should spawn."""
        if len(self._flowers) >= self._config.max_flowers:
            return False

        # If field is completely empty, spawn immediately
        if not self._flowers:
            return True

        self._spawn_accumulator += dt * self._config.spawn_rate * self._config.speed
        if self._spawn_accumulator >= 1.0:
            self._spawn_accumulator -= 1.0
            return True
        return False

    def _generate_flower_position(self) -> tuple[float, float]:
        """Generates organic coordinates with natural spacing."""
        best_x = random.uniform(0.08, 0.92)
        best_y = random.uniform(0.12, 0.88)

        # Attempt to avoid tight clustering if active blooms exist
        if self._flowers:
            max_min_dist = -1.0
            for _ in range(5):
                cand_x = random.uniform(0.08, 0.92)
                cand_y = random.uniform(0.12, 0.88)
                min_d = min(
                    math.hypot(cand_x - f.get_center()[0], cand_y - f.get_center()[1])
                    for f in self._flowers
                )
                if min_d > max_min_dist:
                    max_min_dist = min_d
                    best_x, best_y = cand_x, cand_y

        return best_x, best_y

    def _spawn_flower(self) -> None:
        """Spawns a new organic blooming flower instance."""
        cx, cy = self._generate_flower_position()
        petal_count = random.choice(self._config.petal_options)
        rotation = random.uniform(0.0, 2.0 * math.pi)
        lifespan = random.uniform(2.8, 4.4) / self._config.speed
        max_radius = random.uniform(0.35, 0.52)
        petal_depth = random.uniform(0.18, 0.28)

        # For rainbow palette, generate a fresh randomized harmonious chromatic offset
        if self._config.palette_name.lower() == "rainbow":
            from openrgb_flowers.effects.palettes.rainbow_palette import RainbowPalette
            flower_palette = RainbowPalette(base_hue=random.uniform(0.0, 360.0))
        else:
            flower_palette = self._palette

        flower = FlowerInstance(
            center_x=cx,
            center_y=cy,
            palette=flower_palette,
            petal_count=petal_count,
            max_radius=max_radius,
            lifespan=lifespan,
            rotation=rotation,
            petal_depth=petal_depth,
            spatial_grid=self._spatial_grid,
        )
        self._flowers.append(flower)

    def _compute_ambient_background(self) -> np.ndarray:
        """Calculates subtle gentle meadow ambient wave under keys."""
        if not self._config.ambient_pulse:
            return self._bg_base.copy()

        # Slow breathing botanical breeze across keys
        phase = self._sim_time * 0.8
        wave = (np.sin(self._x_coords * 3.0 + phase) * 0.5 + 0.5)[:, np.newaxis]
        breeze_tint = np.array([6.0, 14.0, 10.0], dtype=np.float32)
        return self._bg_base + breeze_tint * (wave * 0.5)

    def tick(self, dt: float) -> RenderFrame:
        """Advances simulation by dt seconds and renders the next frame."""
        self._sim_time += dt
        self._frame_count += 1

        # 1. Update existing blooms & remove expired
        self._flowers = [f for f in self._flowers if f.update(dt * self._config.speed)]

        # 2. Spawn new blooms if needed
        if self._should_spawn_flower(dt):
            self._spawn_flower()

        # 3. Ambient background
        base_colors = self._compute_ambient_background()

        # 4. Evaluate each flower layer
        layer_colors: List[np.ndarray] = []
        layer_intensities: List[np.ndarray] = []

        for flower in self._flowers:
            intensities, colors = flower.evaluate(self._x_coords, self._y_coords)
            layer_colors.append(colors)
            layer_intensities.append(intensities)

        # 5. Blend layers
        blended = self._blend_strategy.blend_layers(base_colors, layer_colors, layer_intensities)

        # 6. Apply brightness and gamma correction
        final_uint8 = FastColorMath.apply_brightness_gamma(
            blended.astype(np.float32),
            brightness=self._config.brightness,
            gamma=self._config.gamma,
        )

        return RenderFrame(
            timestamp=self._sim_time,
            frame_index=self._frame_count,
            colors=final_uint8,
            led_count=self._led_count,
        )

File: src/effects/blooming_engine.cpp
#include "openrgb_flowers/effects/blooming_engine.hpp"
#include "openrgb_flowers/effects/palettes/palette_registry.hpp"
#include "openrgb_flowers/effects/palettes/rainbow_palette.hpp"
#include "openrgb_flowers/effects/blending/organic_weighted_blend.hpp"
#include "openrgb_flowers/effects/blending/additive_blend.hpp"
#include "openrgb_flowers/effects/blending/screen_blend.hpp"
#include "openrgb_flowers/math/fast_color.hpp"

#include <algorithm>
#include <cmath>
#include <numbers>

namespace openrgb_flowers::effects {

namespace {

std::shared_ptr<core::interfaces::IBlendStrategy> resolve_blend(const std::string& name) {
    if (name == "additive") return std::make_shared<blending::AdditiveBlend>();
    if (name == "screen") return std::make_shared<blending::ScreenBlend>();
    return std::make_shared<blending::OrganicWeightedBlend>();
}

} // anonymous namespace

BloomingEngine::BloomingEngine(
    core::models::EffectConfig config,
    std::shared_ptr<core::interfaces::ILayoutProvider> layout,
    std::shared_ptr<core::interfaces::IColorPalette> palette,
    std::shared_ptr<core::interfaces::IBlendStrategy> blend,
    math::SpatialGrid grid
) : config_(std::move(config)),
    layout_(std::move(layout)),
    palette_(palette ? std::move(palette) : palettes::PaletteRegistry::get(config_.palette_name)),
    blend_strategy_(blend ? std::move(blend) : resolve_blend(config_.blend_mode)),
    grid_(grid),
    rng_(std::random_device{}()) {}

const core::models::EffectConfig& BloomingEngine::get_config() const {
    return config_;
}

int BloomingEngine::get_active_flower_count() const {
    return static_cast<int>(flowers_.size());
}

void BloomingEngine::update_layout(std::shared_ptr<core::interfaces::ILayoutProvider> layout) {
    layout_ = std::move(layout);
}

void BloomingEngine::update_config(const core::models::EffectConfig& config) {
    bool palette_changed = config.palette_name != config_.palette_name;
    bool blend_changed = config.blend_mode != config_.blend_mode;

    config_ = config;

    if (palette_changed) {
        palette_ = palettes::PaletteRegistry::get(config_.palette_name);
    }
    if (blend_changed) {
        blend_strategy_ = resolve_blend(config_.blend_mode);
    }
}

void BloomingEngine::reset() {
    flowers_.clear();
    frame_count_ = 0;
    sim_time_ = 0.0f;
    spawn_accumulator_ = 0.0f;
}

bool BloomingEngine::should_spawn_flower(float dt) {
    if (static_cast<int>(flowers_.size()) >= config_.max_flowers) {
        return false;
    }
    if (flowers_.empty()) {
        return true;
    }
    spawn_accumulator_ += dt * config_.spawn_rate * config_.speed;
    if (spawn_accumulator_ >= 1.0f) {
        spawn_accumulator_ -= 1.0f;
        return true;
    }
    return false;
}

std::pair<float, float> BloomingEngine::generate_flower_position() {
    std::uniform_real_distribution<float> dist_x(0.08f, 0.92f);
    std::uniform_real_distribution<float> dist_y(0.12f, 0.88f);

    float best_x = dist_x(rng_);
    float best_y = dist_y(rng_);

    if (!flowers_.empty()) {
        float max_min_dist = -1.0f;
        for (int i = 0; i < 5; ++i) {
            float cand_x = dist_x(rng_);
            float cand_y = dist_y(rng_);
            float min_d = 1000.0f;
            for (const auto& f : flowers_) {
                auto [cx, cy] = f.get_center();
                float d = std::hypot(cand_x - cx, cand_y - cy);
                if (d < min_d) {
                    min_d = d;
                }
            }
            if (min_d > max_min_dist) {
                max_min_dist = min_d;
                best_x = cand_x;
                best_y = cand_y;
            }
        }
    }
    return {best_x, best_y};
}

void BloomingEngine::spawn_flower() {
    auto [cx, cy] = generate_flower_position();

    int petal_count = 5;
    if (!config_.petal_options.empty()) {
        std::uniform_int_distribution<size_t> petal_dist(0, config_.petal_options.size() - 1);
        petal_count = config_.petal_options[petal_dist(rng_)];
    }

    std::uniform_real_distribution<float> rot_dist(0.0f, 2.0f * std::numbers::pi_v<float>);
    float rotation = rot_dist(rng_);

    std::uniform_real_distribution<float> life_dist(2.8f, 4.4f);
    float lifespan = life_dist(rng_) / config_.speed;

    std::uniform_real_distribution<float> rad_dist(0.35f, 0.52f);
    float max_radius = rad_dist(rng_);

    std::uniform_real_distribution<float> depth_dist(0.18f, 0.28f);
    float petal_depth = depth_dist(rng_);

    std::shared_ptr<core::interfaces::IColorPalette> flower_palette = palette_;
    if (config_.palette_name == "rainbow") {
        std::uniform_real_distribution<float> hue_dist(0.0f, 360.0f);
        flower_palette = std::make_shared<palettes::RainbowPalette>(hue_dist(rng_));
    }

    flowers_.emplace_back(
        cx, cy, flower_palette, petal_count, max_radius, lifespan, rotation, petal_depth, grid_
    );
}

void BloomingEngine::compute_ambient_background(std::vector<core::models::ColorRGB>& out_bg) const {
    size_t key_count = layout_->get_key_count();
    out_bg.resize(key_count);

    if (!config_.ambient_pulse) {
        std::fill(out_bg.begin(), out_bg.end(), config_.background_color);
        return;
    }

    const float* x_coords = layout_->get_x_coords();
    float phase = sim_time_ * 0.8f;
    float bg_r = static_cast<float>(config_.background_color.r);
    float bg_g = static_cast<float>(config_.background_color.g);
    float bg_b = static_cast<float>(config_.background_color.b);

    for (size_t i = 0; i < key_count; ++i) {
        float wave = std::sin(x_coords[i] * 3.0f + phase) * 0.5f + 0.5f;
        float factor = wave * 0.5f;
        uint8_t r = static_cast<uint8_t>(std::clamp(std::round(bg_r + 6.0f * factor), 0.0f, 255.0f));
        uint8_t g = static_cast<uint8_t>(std::clamp(std::round(bg_g + 14.0f * factor), 0.0f, 255.0f));
        uint8_t b = static_cast<uint8_t>(std::clamp(std::round(bg_b + 10.0f * factor), 0.0f, 255.0f));
        out_bg[i] = core::models::ColorRGB(r, g, b);
    }
}

core::models::RenderFrame BloomingEngine::tick(float dt) {
    sim_time_ += dt;
    ++frame_count_;

    // 1. Advance flower lifecycle and purge finished in-place
    std::erase_if(flowers_, [dt_speed = dt * config_.speed](FlowerInstance& f) {
        return !f.update(dt_speed);
    });

    // 2. Spawn new flower if needed
    if (should_spawn_flower(dt)) {
        spawn_flower();
    }

    size_t key_count = layout_->get_key_count();
    const float* x_coords = layout_->get_x_coords();
    const float* y_coords = layout_->get_y_coords();

    // 3. Ambient background
    compute_ambient_background(bg_colors_);

    // 4. Evaluate each active bloom layer
    size_t num_flowers = flowers_.size();
    layer_colors_.resize(num_flowers);
    layer_intensities_.resize(num_flowers);
    for (size_t l = 0; l < num_flowers; ++l) {
        if (layer_colors_[l].size() != key_count) {
            layer_colors_[l].resize(key_count);
            layer_intensities_[l].resize(key_count);
        }
        flowers_[l].evaluate(
            x_coords, y_coords,
            layer_intensities_[l].data(),
            layer_colors_[l].data(),
            key_count
        );
    }

    // 5. Blend layers
    if (blended_.size() != key_count) {
        blended_.resize(key_count);
    }
    blend_strategy_->blend_layers(
        bg_colors_.data(),
        layer_colors_,
        layer_intensities_,
        blended_.data(),
        key_count
    );

    // 6. Post-processing (brightness, saturation, gamma)
    if (final_frame_.size() != key_count) {
        final_frame_.resize(key_count);
    }
    math::FastColorMath::apply_brightness_gamma_buffer(
        blended_.data(),
        final_frame_.data(),
        key_count,
        config_.brightness,
        config_.gamma,
        config_.saturation
    );

    return core::models::RenderFrame(sim_time_, frame_count_, final_frame_);
}

} // namespace openrgb_flowers::effects
